authentication/authentication.go

```diff
@@ -0,0 +1,61 @@
+package authentication
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/auth0/go-auth0/internal/client"
+)
+
+// Authentication is the auth client.
+type Authentication struct {
+	Database *Database
+	OAuth    *OAuth
+
+	url             *url.URL
+	basePath        string
+	debug           bool
+	http            *http.Client
+	auth0ClientInfo *client.Auth0ClientInfo
+	common          manager
+	clientID        string
+	clientSecret    string
+}
+
+type manager struct {
+	authentication *Authentication
+}
+
+// New creates an auth client.
+func New(domain string, options ...Option) (*Authentication, error) {
+	// Ignore the scheme if it was defined in the domain variable, then prefix
+	// with https as it's the only scheme supported by the Auth0 API.
+	if i := strings.Index(domain, "//"); i != -1 {
+		domain = domain[i+2:]
+	}
+	domain = "https://" + domain
+
+	u, err := url.Parse(domain)
+	if err != nil {
+		return nil, err
+	}
+
+	a := &Authentication{
+		url:             u,
+		basePath:        "",
+		debug:           false,
+		http:            http.DefaultClient,
+		auth0ClientInfo: client.DefaultAuth0ClientInfo,
+	}
+
+	for _, option := range options {
+		option(a)
+	}
+
+	a.common.authentication = a
```

Comment:
Could you please expand on the rationale for having a property with a manager that points to the same object?

Reply:
This mirrors a change that was made in the management side as part of #214 (with some inspiration from go-github), the change had 2 main reasons:

1) Prevents the chaining issue that embedding the authentication/management structs introduces (i.e. `authAPI.Database.OAuth.Databse.SignUp()`
2) A pretty small optimisation from reusing a single authentication struct allocation rather one for each manager

Reply:
Thanks for the explanation.